src/core/i18n/translate.js

```diff
@@ -49,6 +44,12 @@ export default function translate(options = {}) {
 
     Translate.displayName = `Translate[${getDisplayName(WrappedComponent)}]`;
 
-    return Translate;
+    function mapStateToProps({ api, i18n }) {
+      return {
+        i18n: makeI18n(i18n, api.lang),
```

Comment:
I think you're right. The code I offered isn't any different than yours. At first I thought that we should prevent `makeI18n()` from getting called on *every* state change regardless of whether the lang state changes or not. Neither my code nor your code prevents that (I think?). Well, it's probably a premature optimization anyway.

If we measure and decide we do need to optimize this, we can probably use the [reselect approach](http://redux.js.org/docs/recipes/ComputingDerivedData.html) with a key on `lang`.